storm-core/src/clj/backtype/storm/daemon/supervisor.clj

```diff
@@ -372,66 +389,50 @@
-                             (pr-str assignment)
-                             " for this supervisor "
-                             (:supervisor-id supervisor)
-                             " on port "
-                             port
-                             " with id "
-                             id
-                             )
-                (launch-worker supervisor
-                               (:storm-id assignment)
-                               port
-                               id
-                               mem-onheap)
-                (mark! supervisor:num-workers-launched)
-                (catch java.io.FileNotFoundException e
-                  (log-message "Unable to launch worker due to "
-                               (.getMessage e)))
-                (catch java.io.IOException e
-                  (log-message "Unable to launch worker due to "
-                               (.getMessage e))))
-         id)))
-    ))
+    (let [valid-new-worker-ids
+          (into {}
+            (remove nil?
+              (dofor [[port assignment] reassign-executors]
+                (let [id (new-worker-ids port)
+                      storm-id (:storm-id assignment)
+                      ^WorkerResources resources (:resources assignment)
+                      mem-onheap (.get_mem_on_heap resources)]
+                  ;; This condition checks for required files exist before launching the worker
+                  (if (required-topo-files-exist? conf storm-id)
+                    (do
+                      (log-message "Launching worker with assignment "
+                        (pr-str assignment)
+                        " for this supervisor "
+                        (:supervisor-id supervisor)
+                        " on port "
+                        port
+                        " with id "
+                        id)
+                      (local-mkdirs (worker-pids-root conf id))
+                      (local-mkdirs (worker-heartbeats-root conf id))
+                      (launch-worker supervisor
+                        (:storm-id assignment)
+                        port
+                        id
+                        mem-onheap)
+                      [port id])
+                    (do
+                      (log-message "Missing topology storm code, so can't launch worker with assignment "
+                        (pr-str assignment)
+                        " for this supervisor "
+                        (:supervisor-id supervisor)
+                        " on port "
+                        port
+                        " with id "
+                        id)
+                      nil))))))]
+      (ls-approved-workers! local-state
+                        (merge
+                          (select-keys (ls-approved-workers local-state)
+                            (keys keepers))
+                          (zipmap (vals valid-new-worker-ids) (keys valid-new-worker-ids))))
```

Comment:
* Some of the indentation looks odd here.
* See [above comment](https://github.com/apache/storm/pull/845/files#r45132995) on avoiding map reversal.